java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,255 @@
+
+  private final Map<Tree, SymbolicValue> map = new HashMap<>();
+  /**
+   * Map to handle lookup of fields.
+   * */
+  private final Map<Symbol, SymbolicValue> symbolMap = new HashMap<>();
+
+  public SymbolicValue createSymbolicValue(Tree syntaxNode) {
+    SymbolicValue result = map.get(syntaxNode);
+    if (result == null && syntaxNode.is(Tree.Kind.IDENTIFIER)) {
+      result = symbolMap.get(((IdentifierTree) syntaxNode).symbol());
+    } else if (result == null && syntaxNode.is(Tree.Kind.VARIABLE)) {
+      result = symbolMap.get(((VariableTree) syntaxNode).symbol());
+    }
+    if (result == null) {
+      result = new SymbolicValue.ObjectSymbolicValue(map.size() + ProgramState.EMPTY_STATE.constraints.size());
+      map.put(syntaxNode, result);
+      if (syntaxNode.is(Tree.Kind.IDENTIFIER)) {
+        symbolMap.put(((IdentifierTree) syntaxNode).symbol(), result);
+      } else if (syntaxNode.is(Tree.Kind.VARIABLE)) {
+        symbolMap.put(((VariableTree) syntaxNode).symbol(), result);
+      }
+    }
+    return result;
+  }
+
+  public SymbolicValue supersedeSymbolicValue(VariableTree variable) {
+    SymbolicValue result = new SymbolicValue.ObjectSymbolicValue(map.size() + ProgramState.EMPTY_STATE.constraints.size());
+    map.put(variable, result);
+    symbolMap.put(variable.symbol(), result);
+    return result;
+  }
+
+  public SymbolicValue eval(ProgramState programState, Tree node) {
+    Tree syntaxNode = skipTrivial(node);
+    switch (syntaxNode.kind()) {
+      case NULL_LITERAL:
+        return SymbolicValue.NULL_LITERAL;
+      case BOOLEAN_LITERAL:
+        boolean value = Boolean.parseBoolean(((LiteralTree) syntaxNode).value());
+        if (value) {
+          return SymbolicValue.TRUE_LITERAL;
+        }
+        return SymbolicValue.FALSE_LITERAL;
+      case VARIABLE:
+        SymbolicValue variableResult = programState.values.get(((VariableTree) syntaxNode).symbol());
+        if (variableResult != null) {
+          // symbolic value associated with local variable
+          return variableResult;
+        }
+        break;
+      case IDENTIFIER:
+        SymbolicValue identifierResult = programState.values.get(((IdentifierTree) syntaxNode).symbol());
+        if (identifierResult != null) {
+          // symbolic value associated with local variable
+          return identifierResult;
+        }
+        break;
+      case LOGICAL_COMPLEMENT:
+        UnaryExpressionTree unaryExpressionTree = (UnaryExpressionTree) syntaxNode;
+        SymbolicValue val = eval(programState, unaryExpressionTree.expression());
+        if (SymbolicValue.FALSE_LITERAL.equals(val)) {
+          return SymbolicValue.TRUE_LITERAL;
+        } else if (val.equals(SymbolicValue.TRUE_LITERAL)) {
+          return SymbolicValue.FALSE_LITERAL;
+        }
+        // if not tied to a concrete value, create symbolic value with no constraint for now.
+        // TODO : create constraint between expression and created symbolic value
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)